### DISCUSSION TITLE
Should src/ be renamed packages/?

### DISCUSSION THREAD

**Discussion by maresb** [pyscaffold/pyscaffold#443] [Ideas]
One of my favorite design decisions about PyScaffold is the `src/` directory. I find it so confusing when one has `projectname/projectname/`. But I have recently come to the conclusion that `src/` is a misnomer.

My original interpretation was that `src/` contains source code. To first order, this is correct, and I find it very useful for distinguishing between `docs/` or `notebooks/`.

My current interpretation is that `src/` is the directory containing top-level package(s) for the project. There are some subtle but very real differences between the two. Here are two examples.

1. Let's suppose you want to include some data file with your project. If you believe that `src/` is for source code, then *data is not source code* and hence does *not* belong in the `src/` directory. Unfortunately I can say from experience that it's not so fun trying to convince setuptools to package a file which is not under `src/`.

2. Let's suppose that you're not an expert on the intricacies of Python's import mechanism. If you have `src/package1/` and `src/package2/`, then I believe it's perfectly reasonable to expect a relative import between `package1` and `package2` to work. Instead, one gets the dreaded "relative import beyond top-level package" error.

With a `packages/` directory (or `pkgs/` or `pkg/` or whatever), I think it would be far more clear that it's acceptable to include data files which are meant for packaging. Also it would help conceptualize that `src/` is not itself a top-level root, but rather each directory therein is a named root.

---

Note: 

I do realize that the FAQ says
> having a dedicated `src` directory to store the package files…

and
> PyScaffold considers the `src` directory to be exclusively dedicated to store files meant to be distributed…

I agree with this phrasing. But it seems to me like the name does not match the function.

---

I'm very curious what others think. This is not (yet) meant to be a proposal. Does it make any sense what I'm saying? Thanks for reading.

**Comment by abravalheri** [comment]
Hi @maresb. I agree that the `src` naming can be a bit confusing...
But I think this is another one of those cases where a historical nomenclature is kept because it ended up becoming a de-facto standard (like the `etc` folder in the UNIX file hierarchy).

My impression is that in the Python ecosystem, this form of organising the project is so popular that:
- experienced developers actually understand when you simply say your project uses the `src`-layout
- you can find a [certain number of articles and blog posts](https://www.google.com/search?q=python+src+layout) published on the topic
- it shows up in the [Python packaging guides](https://packaging.python.org/tutorials/packaging-projects/?highlight=src#creating-the-package-files)
-  and we can argue that `pytest` even [incentivises that for some cases](https://docs.pytest.org/en/stable/goodpractices.html).

By following this convention that pervades the Python community, we make it easy for users to find documentation/tutorials about project organisation or even solutions for common issues, which is a good advantage.

**Comment by FlorianWilhelm** [comment]
Hi @maresb, I also totally agree with your observations and especially with your 2 points. Especially number 1 is often a point of confusion as people see `src` as a bad place for their data files. 

Looking back, `pkgs` might have been a much better name for what the directory actually does. But I also agree with the points of @abravalheri that changing it now might be problematic. What do you think about making a blog post about your observations and points to start off a discussion within the Python community? Maybe it sticks and your idea gets some momentum that will eventually lead to some change.

**Comment by maresb** [comment]
Thanks @abravalheri and @FlorianWilhelm for the feedback.

@abravalheri, it saddens me probably more than it should, that the world is full of terrible conventions which have unfortunately become essential for communication. I notice that I'm desensitized to `/etc/`, but I have the impression that such bad names raise barriers for newcomers. (I'm also reminded of the definition of π as half of a mathematically relevant quantity, but that's a whole other can of worms...)

In my very limited experience, the folks at bugs.python.org are [very receptive to feedback about naming problems](https://bugs.python.org/issue17005#msg370488). It looks like the "Python Packaging User Guide" is managed by PyPA. [This](https://discuss.python.org/c/packaging/14) seems to be their discussion forum. Perhaps that would be a more appropriate venue to raise this?

@FlorianWilhelm I like your idea about a blog post. I'm unfortunately not so social-media savvy. Did you have any particular platform in mind? Innovex was great for your JupyterLab post. My employer m²hycon is in the process of launching a blog at the moment, but it's not yet public. I've been meaning to set up a blog on my personal website, but there's so much other stuff which I need to fix first. Medium seems to have benefits, but I'm very skeptical of their whole model. Since you're in agreement, perhaps it would make sense to coauthor something?

**Reply by abravalheri** [reply]
I completely get you @maresb, good nomenclature is something that if not made properly can be very misleading.

I like the idea of taking this discussion to the broader Python community, and the PyPA forum can be right place for that. Depending on the feedback we can even consider adding an experimental/optional extension supporting the use case.

Regarding a place for publishing, there is also `dev.to`, I used it a few times...

**Reply by abravalheri** [reply]
Going back to your original proposal, I like `pkg`. `pkgs` could also be a bit misleading: as far as I know each package needs its own `setup.py/setup.cfg/pyproject.toml` and therefore it would be trick to offer a template with "multipackage" support.

I noticed that in other communities, like [Ruby](https://www.rubyguides.com/2018/09/ruby-gems-gemfiles-bundler/), [Elixir](https://ieftimov.com/post/writing-elixir-library/) and [Lua](http://lua-users.org/wiki/LuaPackagingGuide), the `lib` folder is also very popular.

**Reply by maresb** [reply]
@abravalheri thanks for the tips. I can't claim to understand the entire setuptools/pip process myself, but I believe that just a single `setup.py/setup.cfg/pyproject.toml` is sufficient.

My perhaps oversimplifed picture is as follows. A "project" consists of a single `setup.py/setup.cfg/pyproject.toml`, typically in the root of a Git repo. (I'm not sure whether or not "project" is an actual Python term, but at least the word appears within `pyproject.toml`.) Then you specify a `package_dir` [where packages can be found](https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html#using-a-src-layout). Either packages under `package_dir` are discovered automatically with `packages=find:` or they can be listed explicitly. (Oddly `package_dir` is a dictionary, but I have not understood why.)

So basically a "project" is a git repo with a single `setup.py/setup.cfg/pyproject.toml` in its root and one or more Python packages under `src/` or `pkg/`.

**Reply by FlorianWilhelm** [reply]
@maresb, regarding the platforms, I think a personal blog or [dev.to](https://dev.to/) as @abravalheri suggests might be the best. In a way, your audience will be hardcore developers and Python maintainers so Medium is quite surely the wrong place for something like that. 

I like your idea. If we all agree here, the three of us working together on a post with you in the lead sounds cool to me. Time is right now also an issue for me as I got some deadlines coming up but I would support as much as I can. When we have the article we could later decide where to post it.

My understanding also was that you can put as many packages under `src` as you want `packages=find:` would add all those to the same Python distribution but I never tried it though. In this case, the plural `pkgs` would look better. It would also be consistent with `tests`, `docs`, `configs` (in dsproject).

**Comment by julie777** [comment]
I don't think src means source code in this case. I was under the impression that it means the source for the package that will be produced by building the project. I could be wrong. 
I am however against using package in yet another way. Python already has a problem that it calls a module that is composed of a directory and multiple files a package. This is confusing for all newcomers who get the obvious definition of package as  the "thing" you install from pypi.

**Comment by FlorianWilhelm** [comment]
There are several good points made here:

* the terms `packages` and `distributions` are clearly defined in the Python ecosystem but often mixed up. Scikit-Learn is the distribution you pip install and `sklearn` is the package. Thus a `pkgs` directory could be misleading as people might think that the whole project somehow makes up the package/distribution,
* people might find it counter-intuitive to put data-files into a `src` folder, thus a `pkgs` folder might be more intuitive,
* the established standard is `src` right now or even no subdirectory at all. One of the goals of Pyscaffold is to somehow reflect the consents of the Python packaging community what is a good practice but it is always a compromise.

So if this was a green field decision, I would vote for `pkgs` or `pkg` as this clearly describes its content, the actual package or packages. @abravalheri, correct me if I am wrong here but I thought that a single distribution could potentially ship several packages, right? Don't know if this ever makes sense. The plural `pkgs` would be more consistent with all the other folders `tests`, `docs`, etc. but in 99.9% of all cases `pkgs` would contain only a single folder, so that might be misleading. 

As `src` is a quasi standard, I am a bit hesitant as it might lead to more confusion. This is why I think it might be helpful to have a blogpost about this or a discussion with the PyPA to get their input too. Was this blog post ever written?

**Reply by abravalheri** [reply]
> I thought that a single distribution could potentially ship several packages, right?

Exactly, one distribution can contain multiple packages (that is the case of [`setuptools-60.8.2-py3-none-any.whl`](https://pypi.org/project/setuptools/60.8.2/#files) for example, that contains `setuptools` and `pkg_resources`).

It is unfortunate that the terminology ended up evolving to be confusing along the years. It was recently brought to my attention that PyPA maintains a glossary in: https://packaging.python.org/en/latest/glossary/.

**Reply by maresb** [reply]
Unfortunately I never got around to writing that blog post (partly due to not having a clear choice of platform, and also wondering if I should raise it directly with PyPA, and finally feeling a bit shy to pursue the issue high-level on a subject where I am relatively new...) I would still really like to see something happen though.

Great find @abravalheri with the glossary.

I wonder if it would be more constructive rather than trying to advocate for renaming `src/` to instead approach things as: "These are the false friend words in PyPA terminology... This is what a reasonable person would expect they mean, but here's what they actually mean."

BTW, It looks to me like PyScaffold hard-codes the `src/` default, right? Or is there some possibility to adjust it?

**Comment by julie777** [comment]
My two cents, and really a sidebar...

Although PyPA defined distribution to address the confusion I have significant issues with that decision and plan to try and get them to reconsider. I think it conflicts with common use and makes the python ecosystem more confusing. 

I install a Linux distribution such as Ubuntu or RedHat and then I install packages using a package manager for things I want to use.  This is exactly how Python works. I install the things I want to use in Python t using a package manager from a repository (pypi), so they should be called packages. 

The PyPA solution was to avoid the real problem (packages/modules) and just make up a new name that doesn't make sense to the majority of users, especially given that Linux and Linux package managers are ubiquitous now.

**Reply by maresb** [reply]
@julie777 It looks like we were typing simultaneously, and you beat me to posting. So this message addresses your idea about complaining to PyPA about their definition of distribution, not about the meaning of `src/`... So apologies that this current message ignores the points in your latest message.

@julie777, I agree with your sentiments, but I wonder if we have any proposal which PyPA could take seriously? This terminology with "package" and "distribution" has already propagated throughout the community, so it's not like PyPA can simply change it based on a few complaints from PyPA outsiders.

Python has been around for many years, and hence PyPA must deal with a huge amount of baggage from previous decisions. I doubt that PyPA just plain sucks at naming things, and they made their decisions too hastily. It's probably more likely that many years ago, something was named suboptimally, and over the years things have snowballed as a result of least-confusing naming decisions into the current mess. But I'm just speculating since I haven't been seriously programming in Python for that many years.

If we want to be taken seriously by PyPA, we'd need a concrete, bulletproof proposal which we could defend against all criticisms. The current names certainly suck, but I feel doubtful that much can be done to fix them at this point.

**Comment by julie777** [comment]
I just copied my conclusion from the end so you don't have to read then entire post.

### Conclusion

_It just occurred to me that after writing down all my thoughts, that the src directory is, using the python path reasoning, just the project equivalent of site-packages. **Maybe src should be named project-packages**._

Thinking ### about it out loud

Looking at the PyScaffold codebase, I notice that there is a egg-info directory under src, which is not source code. 
@maresb I think when you wrote about finding it confusing to have projectname/projectname there is a slight confusion. I used to really prefer the no src directory, but now I see some compelling reasons for having it regarding testing. But the original structure was really project_name/package_name, and those names could have been different if desired. 

I'm sure what the correct name should be for the src directory. I didn't like it years ago when doing C++ development. But the more I think about this the more I support a different name.

Let's see if I can describe it's actually reason for being. When installing a package/module for use it normally is put in site_packages. Site_packages is in the python path and therefore I can `import mypackage` and it works. Without a directory to hold the package/module in a Python dev environement I have to put the project directory in the python path, which allows access to many files and directories that aren't in the package(s) under development. This is not a good idea. 
So:
1. there needs to be a directory that can be put in the python path to all importing code under development
2. when developing a pypi package for upload it is very conceivable that it might contain multiple modules, directory modules(packages), or aggregating modules. They may or may not be namespaced.
3. The pypi package may contain metadat, data files, etc. that are part of the uploaded package that gets built. It isn't source code, but it is part of the pypi package.

I know that I am making this a larger discussion, but sometimes it helps to look at a bigger picture to make the smaller parts clear.

If I start with a project, which maps to a github project/repo then in my project I have project metadata files and directories for different parts of the project. Project metadata is always at the root, although this is just common practice. (There could have been a project_info subdirectory by convention just as easily.)
I typically (always with PyScaffold) have a docs directory for the project, a test directory, a src directory. Others could be supported in the future. 

The project name, the pypi package name, and the module names for the modules used by the users are not intrinsically linked. They could all be different, although most of the time they are the same because it is simple and avoids confusion. If we look at a pypi package like boltons, we see that the package name boltsons is really just an aggregating namespace for all of the individual modules that it contains. It also provides an aggregate package name for pypi. I haven't look for the source project on github, but I suspect that it shares the boltons name for convenience. On the other hand I can point to just as many github projects that have a name, but that isn't what the user actually installs or uses.

That leaves me with a question for those of you who have been building and distributing python packages much longer than I (as most of my work has been in organizations where I was building applications and deploying them).
Does the src directory include everything that should get put in the pypi package, or is it just the python module/package part of the pypi package?

This is a chance to be clear about what directories in a python project actually mean. 
- doc - this is the source for what gets uploaded to the doc site
- test - this is part of the project, but never distributed
- src - this is the source for the pypi package
etc.

This leads me to more questions.
Is the project README displayed by github the same documentation for pypi?
Is the license that is project metadat put in the pypi package?
What about other metadata files. 
**Is it possible to upload a package to pypi that is multiple separate modules, or does it always require aggregating into a single namespace package?**

At the end of all my thinking out loud, I really think that distributed chunks, which have different destinations, such as doc, the pypi package are artifacts of the build and don't have an exact correspondence to project directory structure and don't need to. A single file in the project might be used as part of building any number of artifacts.

Given that the only real function of the src directory is to keep from putting the project directory in the python path, I believe we can name it anything that makes the most sense to the most people.

One more note from pip
> python -m pip install -e path/to/SomeProject
really isn't correct, as it really means path/to/directory_containing_modules_that_I_want_to_import

I have never experienced the problems that are used as the reason for a src directory so I am still on the fence about it all.

Sorry to ramble. (I am just trying to contribute to the discussion.)

It just occurred to me that after writing all that what the src directory really is, using the python path reasoning, is the project equivalent of site-packages. _Maybe src should be named project-packages._

**Reply by maresb** [reply]
@julie777, your thought process is very similar to mine.

To hastily answer some of your questions off the top of my head (correctness not verified):

> Does the src directory include everything that should get put in the pypi package, or is it just the python module/package part of the pypi package?

The PyPI package typically also includes metadata-type stuff like LICENSE, and README, as well as the `setup.py` installation script (when using `setuptools`). Included files are traditionally determined by `MANIFEST.in`, or in the case of PyScaffold by `setuptools-scm`, or possibly by whatever newfangled `pyproject.toml` stuff which is being introduced. So it's more the module/package part. I'd encourage you to look inside some sdists and wheels to get a feel for things. (Wheels are `.zip`s with a `.whl` extension.)

> Is the project README displayed by github the same documentation for pypi?

Roughly yes... by convention GitHub displays whatever README may exist in the project root. When a project is built into a distribution for PyPI, such a README is typically copied into the distribution. PyPI simply reads whatever distribution is uploaded, and displays any corresponding README. So in practice, the README displayed by PyPI is the state of the README from GitHub at the point in time of the previous release. (Consequently, to update a README on PyPI you must perform an entire release.)

> Is the license that is project metadat put in the pypi package?

The dist builder (e.g. `setuptools`) contains the logic for exactly what to do here. Anything can happen, but typically the LICENSE ends up getting bundled. But not always, and this can lead to a mess where your distribution lacks a license, creating a legal hurdle which prevents, for example, repackaging on Conda-Forge.

> Is it possible to upload a package to pypi that is multiple separate modules, or does it always require aggregating into a single namespace package?

A single build/distribution from PyPI can install any number of separate modules/packages.

> Maybe src should be named project-packages.

That would be sensible. My impression is that the `project-` part may be a redundant since of course it's part of your project. But I haven't seen anything which I would consider the "perfect name". Naming things is hard! :joy:

**Comment by julie777** [comment]
@maresb 
I only added the project- prefix to be consistent with site-packages. 

Now I realize that although site-packages originally meant packages installed for the site, much like /etc/<config> vs HOME/<config>, it has changed and there are many places that a directory named site-packages will be created. If I install Python as a user, it will have a site-packages, but that is really user-packages. I'm sure you get the point.

Another thing that i have realized, after reading at least 100 pages of PyPA related documentation is that 
1. the name src for that directory is already very entrenched in the Python ecosystem and we should play nice and use it too.
2. It really is src and not just packages. The way setuptools works is that anything you want in your distribution should be in src. That allow setuptools to make your life easy. You don't have to create a Manifest file. You create as many non-package sub-directories under src as you need. 
 
The extra level provided for source does two things.
1. allow adding src to python path so that packages and modules can be found running in dev mode
2. provide a grouping of all files that are part of the distribution so adding them during packaging is automated

**Comment by maresb** [comment]
> It really is src and not just packages. The way setuptools works is that anything you want in your distribution should be in src.

@julie777, are you implying that `src/` is hard-coded in setuptools? As far as I know this is not the case, and it must be specified in the configuration, for example [here](https://github.com/pyscaffold/pyscaffold-demo/blob/812eef9c1da6f7e2d24d96ccb0c302b0e8f471bb/setup.cfg#L39).

**Comment by julie777** [comment]
@maresb 

No, I wasn't implying it was hard coded in any tools but rather that it is on numerous pages of documentation by the group (PyPA) with the most clout for standardization. 

In addition there are many write-ups by various people about reasons why using src is a good idea.

My personal opinion, after much thinking about what you said and doing my own research,  is that I like src because it is the source that is used to create the package. But, I am just one voice. 😄
